test_game.py

```diff
@@ -0,0 +1,150 @@
+    enemy_moves = []
+    my_moves = []
+
+    def __init__(self):
+        pass
+
+    def generate_move(self):
+        """@brief When this player must make a move, this function generates it."""
+        # This logic should be more advanced.
+        selected_move = MoveInfo(public_symbol=0, private_symbol=0)
+
+        self.my_moves.append(selected_move)
+        return selected_move
+
+    def be_advised(self, enemy_move):
+        """@brief After an enemy turn, this player is informed at least partially of the enemy's action."""
+        self.enemy_moves.append(enemy_move)
+
+
+class TestGame:
+    """The class that contains the main game loop."""
+    rounds = 0
+    players = []  # Set of player classes
+    moves = []  # Set of sets of moves, one set of moves per players
+
+    def __init__(self, players, rounds=1):
+        self.players = players
+        self.rounds = rounds
+
+        for _ in players:
+            self.moves.append([])
+
+    def game_loop(self):
+        """@brief The game's main loop."""
+        for round_number in range(0, self.rounds):
+            player_number = round_number % len(self.players)
+
+            players_move = self.players[player_number].generate_move()  # The player generates the next move.
+            self.moves[player_number].append(players_move)  # The player's move is recorded into the game history
+
+            # The player's move is processed to determine how every OTHER player views it, and then every other player
+            #     is informed about the move. In this case, there is only one player, which always sees the previous
+            #     move in exactly one way. However, with more players, this logic may get more complex.
+            last_shroud = ShroudedMoveInfo()
+            last_shroud.public_symbol = players_move.public_symbol
+
+            # Inform all other players
+            for other_player_number, player in enumerate(self.players):
+                if other_player_number == player_number:
+                    continue  # Don't advise ourselves
+
+                player.be_advised(last_shroud)
+
+
+
+        self.conclude()  # End game
+
+
+    def conclude(self):
+        """@brief Final wrap - calculate scores and declare a winner."""
+        scores = []
+        for _ in self.players:
+            scores.append(0)
+
+        for player_number, player_move_set in enumerate(self.moves):  # For each player...
+            for move in player_move_set:  # For each move they made...
+                # Calculate how much it adds to their score.
+                public_symbol = move.public_symbol
+                private_symbol = move.private_symbol
+
+                for other_player_number, other_player_move_set in enumerate(self.moves):
+                    if other_player_number == player_number:
+                        continue  # Don't compare with our own moves
+
+                    for enemy_move in other_player_move_set:
```

Comment:
A few ways to improve the efficiency of these nested loops:
- in the `other_player_number` loop, iterate over enumerate(self.moves[player_number:]) to avoid symmetry
- do some calculations based on the total number of each type of move each player has made, rather than comparing individual moves directly

I guess this is only called once per game though so it doesn't matter all that much for now

Reply:
If it doesn't matter for now, I'd rather leave it until there's a use for it...

Reply:
Sounds good.